pkg/timeseries/extent_list.go

```diff
@@ -445,47 +450,53 @@ func (el ExtentList) TimestampCount(d time.Duration) int64 {
 	return c
 }
 
-// CalculateDeltas provides a list of extents that are not in a cached timeseries,
-// when provided a list of extents that are cached.
-func (el ExtentList) CalculateDeltas(want Extent, step time.Duration) ExtentList {
+// CalculateDeltas provides a list of extents that are not in el based on the
+// needed extent. step is used to determine which absolute timestamps in need
+// will be checked in el.
+func (el ExtentList) CalculateDeltas(need Extent, step time.Duration) ExtentList {
+	if step <= 0 || !need.End.After(need.Start) {
+		return ExtentList{}
+	}
 	if len(el) == 0 {
-		return ExtentList{want}
-	}
-	misCap := want.End.Sub(want.Start) / step
-	if misCap < 0 {
-		misCap = 0
-	}
-	misses := make([]time.Time, 0, misCap)
-	for i := want.Start; !want.End.Before(i); i = i.Add(step) {
-		found := false
-		for j := range el {
-			if j == 0 && i.Before(el[j].Start) {
-				// our earliest datapoint in cache is after the first point the user wants
-				break
+		return ExtentList{need}
+	}
+	sort.Sort(el)
+	out := make(ExtentList, len(el)+1)
+	var missStart time.Time
+	var j, k int
+	for ts := need.Start; !ts.After(need.End); ts = ts.Add(step) {
+		// this advances to j to the point in el where ts would be if it were
```

Comment:
(minor) nit, wording seems off. perhaps:
```suggestion
		// this advances j to the point in el where ts would be if it were
```